<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Mapping and Binning a Metagenome Assembly &mdash; 2017-dibsi-metagenomics 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinxcontrib-images/LightBox2/lightbox2/css/lightbox.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2/js/lightbox.min.js"></script>
    <script type="text/javascript" src="_static/sphinxcontrib-images/LightBox2/lightbox2-customize/jquery-noconflict.js"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.3.6/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="2017-dibsi-metagenomics 1.0 documentation" href="toc.html" />
    <link rel="next" title="Gene Abundance Estimation with Salmon" href="salmon_tutorial.html" />
    <link rel="prev" title="Annotation with Prokka" href="prokka_tutorial.html" />


<style type="text/css">
<!-- github-based editing disabled; set github_base_account
     and github_project -->
#editor-trap.toggled {
  display: none;
}
</style>



  </head>
  <body role="document">

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="toc.html">
          2017-dibsi-metagenomics</a>
        <span class="navbar-text navbar-version pull-left"><b>1.0</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="toc.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">2017 Metagenomics workshop at UC Santa Cruz</a></li>
<li class="toctree-l1"><a class="reference internal" href="welcome.html">Welcome!</a></li>
<li class="toctree-l1"><a class="reference internal" href="jetstream/boot.html">Booting a Jetstream Computer Instance for your use!</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality.html">Short read quality and trimming</a></li>
<li class="toctree-l1"><a class="reference internal" href="assemble.html">Run the MEGAHIT assembler</a></li>
<li class="toctree-l1"><a class="reference internal" href="prokka_tutorial.html">Annotation with Prokka</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Mapping and Binning a Metagenome Assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="salmon_tutorial.html">Gene Abundance Estimation with Salmon</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html">Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="slice.html">Slicing and dicing with k-mers</a></li>
<li class="toctree-l1"><a class="reference internal" href="circos_tutorial.html">Using and Installing Circos</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">Workflow and repeatability discussion</a></li>
<li class="toctree-l1"><a class="reference internal" href="DATA.html">Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="INSTALL.html">Building this site on your own</a></li>
<li class="toctree-l1"><a class="reference internal" href="README.html">2017 / April / Metagenomics</a></li>
<li class="toctree-l1"><a class="reference internal" href="day2-install.html">Day 2 - installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="kmer_trimming.html">K-mer Spectral Error Trimming</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Mapping and Binning a Metagenome Assembly</a><ul>
<li><a class="reference internal" href="#mapping">Mapping</a><ul>
<li><a class="reference internal" href="#downloading-data">Downloading data</a></li>
<li><a class="reference internal" href="#optional-look-at-tetramer-clustering">Optional: look at tetramer clustering</a></li>
<li><a class="reference internal" href="#mapping-the-reads">Mapping the reads</a></li>
<li><a class="reference internal" href="#converting-to-bam-to-quickly-quantify">Converting to BAM to quickly quantify</a></li>
</ul>
</li>
<li><a class="reference internal" href="#binning">Binning</a><ul>
<li><a class="reference internal" href="#binning-the-assembly">Binning the assembly</a></li>
<li><a class="reference internal" href="#visualizing-the-bins">Visualizing the bins</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
  <li>
    <a href="prokka_tutorial.html" title="Previous Chapter: Annotation with Prokka"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; Annotation wi...</span>
    </a>
  </li>
  <li>
    <a href="salmon_tutorial.html" title="Next Chapter: Gene Abundance Estimation with Salmon"><span class="glyphicon glyphicon-chevron-right visible-sm"></span><span class="hidden-sm hidden-tablet">Gene Abundanc... &raquo;</span>
    </a>
  </li>
              
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="_sources/binning.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">

    <div class="col-md-12 content">
      
  
  <div class="section" id="mapping-and-binning-a-metagenome-assembly">
<h1>Mapping and Binning a Metagenome Assembly<a class="headerlink" href="#mapping-and-binning-a-metagenome-assembly" title="Permalink to this headline">¶</a></h1>
<p>A common approach following metagenome assembly is binning, a process by which assembled contigs are collected into groups or &#8216;bins&#8217; that might then be assigned some taxonomic affiliation. There are many different tools that can be used for binning (see <a class="reference external" href="http://biorxiv.org/content/early/2017/01/09/099127">CAMI review for more details</a>). Here, we will be using <a class="reference external" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/2049-2618-2-26">MaxBin</a>, which is both user friendly and highly cited. To use MaxBin, we will first need to map our data against the assembled metagenome using bwa and then estimate relative abundances by contig. We will then inspect the bins generated by MaxBin using VizBin.</p>
<div class="section" id="mapping">
<h2>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h2>
<p>Download bwa:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>cd
curl -L https://sourceforge.net/projects/bio-bwa/files/bwa-0.7.15.tar.bz2/download &gt; bwa-0.7.15.tar.bz2
</pre></div>
</div>
<p>Unpack and build it:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>tar xjvf bwa-0.7.15.tar.bz2
cd bwa-0.7.15
make
</pre></div>
</div>
<p>Install it:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>sudo cp bwa /usr/local/bin
</pre></div>
</div>
<div class="section" id="downloading-data">
<h3>Downloading data<a class="headerlink" href="#downloading-data" title="Permalink to this headline">¶</a></h3>
<p>Now, go to a new directory and grab the data:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>mkdir ~/mapping
cd ~/mapping

curl -O https://s3-us-west-1.amazonaws.com/dib-training.ucdavis.edu/metagenomics-scripps-2016-10-12/SRR1976948.abundtrim.subset.pe.fq.gz
curl -O https://s3-us-west-1.amazonaws.com/dib-training.ucdavis.edu/metagenomics-scripps-2016-10-12/SRR1977249.abundtrim.subset.pe.fq.gz
</pre></div>
</div>
<p>And extract the files:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>for file in *fq.gz
  do
  gunzip $file
done
</pre></div>
</div>
<p>We will also need the assembly; rather than rebuilding it, you can download a copy that we saved for you:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>curl -O https://s3-us-west-1.amazonaws.com/dib-training.ucdavis.edu/metagenomics-scripps-2016-10-12/subset_assembly.fa.gz
gunzip subset_assembly.fa.gz
</pre></div>
</div>
</div>
<div class="section" id="optional-look-at-tetramer-clustering">
<h3>Optional: look at tetramer clustering<a class="headerlink" href="#optional-look-at-tetramer-clustering" title="Permalink to this headline">¶</a></h3>
<p>The mapping is going to take a while, so while it&#8217;s running we can look at
the tetramer nucleotide frequencies of all of our assembled contigs.  For
that, we&#8217;ll need <a class="reference external" href="https://github.com/ngs-docs/2017-ucsc-metagenomics/blob/master/files/sourmash_tetramer.ipynb">this notebook</a>:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>cd ~/
curl -O https://raw.githubusercontent.com/ngs-docs/2017-ucsc-metagenomics/master/files/sourmash_tetramer.ipynb
cd -
</pre></div>
</div>
<p>and during the mapping we&#8217;ll load this notebook in the Jupyter notebook
console.  Note, the address of that will be:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>echo http://$(hostname):8000/
</pre></div>
</div>
<p>...back to mapping!</p>
</div>
<div class="section" id="mapping-the-reads">
<h3>Mapping the reads<a class="headerlink" href="#mapping-the-reads" title="Permalink to this headline">¶</a></h3>
<p>First, we will need to to index the megahit assembly:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>bwa index subset_assembly.fa
</pre></div>
</div>
<p>Now, we can map the reads:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>for i in *fq
  do
      bwa mem -p subset_assembly.fa $i &gt; ${i}.aln.sam
  done
</pre></div>
</div>
</div>
<div class="section" id="converting-to-bam-to-quickly-quantify">
<h3>Converting to BAM to quickly quantify<a class="headerlink" href="#converting-to-bam-to-quickly-quantify" title="Permalink to this headline">¶</a></h3>
<p>First, index the assembly for samtools:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>samtools faidx subset_assembly.fa
</pre></div>
</div>
<p>Then, convert both SAM files to BAM files:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>for i in *.sam
  do
    samtools import subset_assembly.fa $i $i.bam
    samtools sort $i.bam $i.bam.sorted
    samtools index $i.bam.sorted.bam
  done
</pre></div>
</div>
<p>Now we can quickly get a rough count of the total number of reads mapping to each contig using samtools:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>for i in *sorted.bam
  do
    samtools idxstats $i &gt; $i.idxstats
    cut -f1,3 $i.idxstats &gt; $i.counts
  done
</pre></div>
</div>
</div>
</div>
<div class="section" id="binning">
<h2>Binning<a class="headerlink" href="#binning" title="Permalink to this headline">¶</a></h2>
<p>First, install MaxBin:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>cd
curl  https://downloads.jbei.org/data/microbial_communities/MaxBin/getfile.php?MaxBin-2.2.2.tar.gz &gt; MaxBin-2.2.2.tar.gz
tar xzvf MaxBin-2.2.2.tar.gz
cd MaxBin-2.2.2
cd src
make
cd ..
./autobuild_auxiliary
export PATH=$PATH:~/MaxBin-2.2.2
</pre></div>
</div>
<div class="section" id="binning-the-assembly">
<h3>Binning the assembly<a class="headerlink" href="#binning-the-assembly" title="Permalink to this headline">¶</a></h3>
<p>Finally, run the MaxBin! Note: MaxBin can take a lot of time to run and bin your metagenome. As this is a workshop, we are doing three things that sacrifice <em>quality</em> for <em>speed</em>.</p>
<ol class="arabic simple">
<li>We are only using 2 of the 6 datasets that were generated for the
this project. Most binning software, including MaxBin, rely upon
many samples to accurately bin data. And, we have subsampled the
data to make it faster to proess.</li>
<li>We did quick and dirty contig estimation using samtools. If you use MaxBin, I would recommend allowing them to generate the abudance tables using Bowtie2. But, again, this will add to the run time.</li>
<li>We are limiting the number of iterations that are performed through
their expectation-maximization algorithm (5 iterations instead of
50+). This will likely limit the quality of the bins we get
out. So, users beware and read <a class="reference external" href="https://downloads.jbei.org/data/microbial_communities/MaxBin/README.txt">the user&#8217;s manual</a>
before proceeding with your own data analysis.</li>
</ol>
<p>First, we will get a list of the count files that we have to pass to MaxBin:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>cd ~/mapping
mkdir binning
cd binning
ls ../*counts &gt; abundance.list
</pre></div>
</div>
<p>Now, on to the actual binning:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>run_MaxBin.pl -contig ../subset_assembly.fa -abund_list abundance.list -max_iteration 5 -out mbin
</pre></div>
</div>
<p>This will generate a series of files. Take a look at the files generated. In particular you should see a series of <a href="#id1"><span class="problematic" id="id2">*</span></a>.fasta files preceeded by numbers. These are the different genome bins predicted by MaxBin.</p>
<p>Take a look at the mbin.summary file. What is shown?</p>
<p>Now, we are going to generate a concatenated file that contains all of our genome bins put together. We will change the fasta header name to include the bin number so that we can tell them apart later.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>for file in mbin.*.fasta
  do
    num=${file//[!0-9]/}
    sed -e &quot;/^&gt;/ s/$/ ${num}/&quot; mbin.$num.fasta &gt;&gt; binned.concat.fasta
  done
</pre></div>
</div>
<p>And finally make an annotation file for visualization:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>echo label &gt; annotation.list
grep &quot;&gt;&quot; binned.concat.fasta |cut -f2 -d &#39; &#39;&gt;&gt; annotation.list
</pre></div>
</div>
</div>
<div class="section" id="visualizing-the-bins">
<h3>Visualizing the bins<a class="headerlink" href="#visualizing-the-bins" title="Permalink to this headline">¶</a></h3>
<p>Now that we have our binned data there are several different things we can do. One thing we might want to do is check the quality of the binning&#8211; a useful tool for this is CheckM. We can also visualize the bins that we just generated using VizBin.</p>
<p>First, install VizBin:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>cd
sudo apt-get install libatlas3-base libopenblas-base
curl -L https://github.com/claczny/VizBin/blob/master/VizBin-dist.jar?raw=true &gt; VizBin-dist.jar
</pre></div>
</div>
<p>VizBin can run in OSX or Linux but is very hard to install on Windows. To simplify things we are going to run VizBin in the desktop emulator through JetStream (which is ... a bit clunky). So, go back to the Jetstream and open up the web desktop simulator.</p>
<a class=""
               data-lightbox="group-5e2bd80a-1b4d-4123-9f59-75fb84fb90e4"
               href="_images/VizBin-OpenDesktop.png"
               title=""
               data-title=""
               ><img src="_images/VizBin-OpenDesktop.png"
                     class=""
                     width="50%"
                     height="auto"
                     alt=""/>
                </a><p>Open the terminal through the desktop simulator and open VizBin:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>java -jar VizBin-dist.jar
</pre></div>
</div>
<p>This should prompt VizBin to open in another window. Click the choose button to open file browser to navigate to the binning folder (~/mapping/binning). There you will find the concatenated binned fasta file (binned.concat.fasta). Upload this file and hit run.</p>
<a class=""
               data-lightbox="group-f7ac7723-cc59-4f69-bf1e-e306379d7c56"
               href="_images/VizBin-LoadFile.png"
               title=""
               data-title=""
               ><img src="_images/VizBin-LoadFile.png"
                     class=""
                     width="50%"
                     height="auto"
                     alt=""/>
                </a><p>What do you see? Read up a bit on <a class="reference external" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-014-0066-1">VizBin</a> to see how the visualization is generated.</p>
<p>Now, upload the annotation.list file as an annotation file to VizBin. The annotation file contains the bin id for each of the contigs in the assembly that were binned.</p>
<a class=""
               data-lightbox="group-4456bb1d-48ae-4a73-a72e-c5e7c2ac1155"
               href="_images/VizBin-AddFiles.png"
               title=""
               data-title=""
               ><img src="_images/VizBin-AddFiles.png"
                     class=""
                     width="50%"
                     height="auto"
                     alt=""/>
                </a><p>How do the two binning methods look in comparison?</p>
<p>Now we will <a class="reference external" href="https://2017-ucsc-metagenomics.readthedocs.io/en/latest/sourmash.html">search and compare our data</a>.</p>
</div>
</div>
</div>



<hr>

<font size="-1"><b>LICENSE:</b>
This documentation and all textual/graphic site content is released
under <a href='http://creativecommons.org/publicdomain/zero/1.0/'>
Creative Commons - 0
(CC0)</a> -- <a href='https://github.com/ngs-docs/2017-ucsc-metagenomics'>fork @
github</a>. </font>

<hr>


<!-- disqus commenting disabled; set disqus_shortname -->

    </div>
    

  </div>
</div>

<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2017, Lab for Data Intensive Biology.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.5.<br/>
    </p>
  </div>
</footer>




<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', '', 'auto');
  ga('send', 'pageview');

</script>


  </body>
</html>